### PR TITLE
Refactor `Hbc::SystemCommand`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Move `HashValidator` and use `ErrorDuringExecution`.

### CODE REVIEW

File: Library/Homebrew/exceptions.rb [hunk 1]
```diff
@@ -524,9 +526,22 @@ def initialize(cause)
...
+      s << "==> Standard Output of failed command:\n"
+      s << stdout
+      s << "\n" unless stdout.end_with?("\n")
+    end
+
+    if stderr
+      s << "==> Standard Error of failed command:\n"
+      s << stderr
+      s << "\n" unless stderr.end_with?("\n")
+    end
```
Comment (JCount) [inline Library/Homebrew/exceptions.rb]: It probably would be good to expand `ErrorDuringExecution`'s test so that there isn't decrease in coverage here.

**Comment by reitermarkus** [comment]
Changed the output to be interleaved and added some tests.

File: Library/Homebrew/exceptions.rb [hunk 2]
```diff
@@ -524,9 +526,16 @@ def initialize(cause)
...
 class ErrorDuringExecution < RuntimeError
-  def initialize(cmd, args = [])
-    args = args.map { |a| a.to_s.gsub " ", "\\ " }.join(" ")
-    super "Failure while executing: #{cmd} #{args}"
+  def initialize(cmd, status:, output: nil)
+    s = "Failure while executing; `#{cmd.shelljoin.gsub(/\\=/, "=")}` exited with #{status.exitstatus}."
+
+    unless [*output].empty?
+      s << " Here's the output:\n"
+      s << output.map { |(type, line)| (type == :stderr) ? Formatter.error(line) : line }.join
```
Comment (MikeMcQuaid) [inline Library/Homebrew/exceptions.rb]: Can this go into a multiline `map` with an `if`/`else`? It's quite hard to parse as-is. Thanks!
